pkg/services/authz/rbac/service.go

```diff
@@ -210,3 +216,85 @@ func getScopeMap(permissions []accesscontrol.Permission) map[string]bool {
 	}
 	return permMap
 }
+
+func (s *Service) checkInheritedPermissions(ctx context.Context, scopeMap map[string]bool, req *CheckRequest) (bool, error) {
+	if req.ParentFolder == "" {
+		return false, nil
+	}
+
+	ctxLogger := s.logger.FromContext(ctx)
+
+	folderMap, err := s.buildFolderAndDashTree(ctx, req.Namespace)
+	if err != nil {
+		ctxLogger.Error("could not build folder and dashboard tree", "error", err)
+		return false, err
+	}
+
+	currentUID := req.ParentFolder
+	for {
+		if node, has := folderMap[currentUID]; has {
+			scope := dashboards.ScopeFoldersProvider.GetResourceScopeUID(node.uid)
+			if scopeMap[scope] {
+				return true, nil
+			}
+			if node.parentUID == nil {
+				break
+			}
+			currentUID = *node.parentUID
+		} else {
+			break
+		}
+	}
+	return false, nil
+}
+
+func (s *Service) buildFolderAndDashTree(ctx context.Context, ns claims.NamespaceInfo) (map[string]folderNode, error) {
+	folders, err := s.store.GetFolders(ctx, ns)
+	if err != nil {
+		return nil, fmt.Errorf("could not get folders: %w", err)
+	}
+	dashboards, err := s.store.GetDashboards(ctx, ns)
+	if err != nil {
+		return nil, fmt.Errorf("could not get dashboards: %w", err)
+	}
+
+	folderMap := make(map[string]folderNode, len(folders))
+	for _, folder := range folders {
+		if node, has := folderMap[folder.UID]; !has {
+			folderMap[folder.UID] = folderNode{
+				uid:       folder.UID,
+				parentUID: folder.ParentUID,
+			}
+		} else {
+			node.parentUID = folder.ParentUID
+			folderMap[folder.UID] = node
+		}
+		// Register that the parent has this child node
+		if folder.ParentUID == nil {
+			continue
+		}
+		if parent, has := folderMap[*folder.ParentUID]; has {
+			parent.childrenUIDs = append(parent.childrenUIDs, folder.UID)
+			folderMap[*folder.ParentUID] = parent
+		} else {
+			folderMap[*folder.ParentUID] = folderNode{
+				uid:          *folder.ParentUID,
+				childrenUIDs: []string{folder.UID},
+			}
+		}
+	}
+
+	for _, dash := range dashboards {
+		node := folderNode{uid: dash.UID, parentUID: dash.ParentUID}
```

Comment:
I wonder if we should have a way to differentiate folders from dashboards when listing. Although, maybe it's a bit early to consider. But I think the use case would be: list the dashboards the user has <verb> access to. In that case, I think, the folder uids should be filtered out.

Reply:
I think you're right that we expect folders and dashboards to be returned in separate lists when listing. But after George's feedback I've decided not to fetch folders at all, and we can infer the folders that the user has direct permissions on purely from the permissions themselves (as in, by looking at the scope).